### PR TITLE
Monitor selection

### LINKED ISSUE #85: Monitor selection for capture

### PR DESCRIPTION
**Pull request by tataRinaz**
Closes #85

### CODE REVIEW

File: src/app/main.cpp [hunk 1]
```diff
@@ -28,21 +28,36 @@ static CapturePtr create_capture(Context context) {
...
-  auto monitor = sc::GetMonitors().front();
-  auto width   = context.m_config->get<std::size_t>("width", static_cast<std::size_t>(monitor.Width));
-  auto height  = context.m_config->get<std::size_t>("height", static_cast<std::size_t>(monitor.Height));
-  shar::Size frame_size {height, width};
+
+  auto monitor_number = context.m_config->get<std::size_t>("monitor", 0);
+  auto monitors = sc::GetMonitors();
+  if (monitor_number >= monitors.size()) {
+    std::string monitors_info;
+    for (auto monitor : monitors) {
```
Comment (0xd34d10cc) [inline src/app/main.cpp]: iterate by references

File: src/app/main.cpp [hunk 2]
```diff
@@ -28,21 +28,36 @@ static CapturePtr create_capture(Context context) {
...
+      monitors_info += std::string(
+        std::to_string(monitor.Index) + " "
+        + monitor.Name
+        + std::to_string(monitor.Width) + "x"
+        + std::to_string(monitor.Height) + '\n');
+    }
+    context.m_logger.info("Available monitors:\n"+monitors_info);
+    throw std::runtime_error("Selected "+std::to_string(monitor_number)+" monitor unavailable");
+  }
+  auto monitor = monitors[monitor_number];
```
Comment (0xd34d10cc) [inline src/app/main.cpp]: Use reference. No reason to copy.

File: src/app/main.cpp [hunk 3]
```diff
@@ -28,21 +28,36 @@ static CapturePtr create_capture(Context context) {
   auto monitor = sc::GetMonitors().front();
```
Comment (pl0q1n) [inline src/app/main.cpp]: why do you use sc::GetMonitors().front() here. it doesn't work with multi monitors at all. you should check monitors before logging any information about monitor name, width, etc.

File: src/app/main.cpp [hunk 4]
```diff
@@ -67,17 +84,18 @@ static prometheus::Exposer create_exposer(Context& context) {
...
-  return Context{config, logger, metrics};
+  return Context{ config, logger, metrics };
 }
 
 static int run() {
   auto context = make_context();
   auto exposer = create_exposer(context);
-  auto capture = create_capture(context);
-  auto encoder = create_encoder(context);
+  const auto& monitor = select_monitor(context);
```
Comment (0xd34d10cc) [inline src/app/main.cpp]: There is no reason to rely on "const reference lifetime extension" feature in C++17.

File: src/encoder/ffmpeg/codec.cpp [hunk 5]
```diff
@@ -13,6 +13,69 @@ namespace shar::codecs::ffmpeg {
 // Clock rate (number of ticks in 1 second) for H264 video. (RFC 6184 Section 8.2.1)
 static const unsigned int CLOCK_RATE = 90000;
 
+AVCodecContext* Codec::make_context(const ConfigPtr& config, AVCodec* codec, Size frame_size, std::size_t fps) {
```
Comment (0xd34d10cc) [inline src/encoder/ffmpeg/codec.cpp]: Why this code is in the diff?

File: src/encoder/ffmpeg/codec.hpp [hunk 6]
```diff
@@ -26,6 +25,8 @@ class Codec {
   Codec& operator=(Codec&&) = delete;
   ~Codec() = default;
 
+  AVCodecContext* make_context(const ConfigPtr& config, AVCodec* codec, Size frame_size, std::size_t fps);
```
Comment (0xd34d10cc) [inline src/encoder/ffmpeg/codec.hpp]: ...and this?

**Pull request merged**